src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,424 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
   return new LateInlineVectorReboxingCallGenerator(method, inline_cg);
 }
 
+class LateInlineScopedValueCallGenerator : public LateInlineCallGenerator {
+  Node* _sv;
+  bool _process_result;
+
+public:
+  LateInlineScopedValueCallGenerator(ciMethod* method, CallGenerator* inline_cg, bool process_result) :
+          LateInlineCallGenerator(method, inline_cg), _sv(nullptr), _process_result(process_result) {}
+
+  virtual JVMState* generate(JVMState* jvms) {
+    Compile *C = Compile::current();
+
+    C->log_inline_id(this);
+
+    C->add_scoped_value_late_inline(this);
+
+    JVMState* new_jvms = DirectCallGenerator::generate(jvms);
+    return new_jvms;
+  }
+
+  virtual CallGenerator* with_call_node(CallNode* call) {
+    LateInlineScopedValueCallGenerator* cg = new LateInlineScopedValueCallGenerator(method(), _inline_cg, false);
+    cg->set_call_node(call->as_CallStaticJava());
+    return cg;
+  }
+
+  void do_late_inline() {
+    CallNode* call = call_node();
+    _sv = call->in(TypeFunc::Parms);
+    CallGenerator::do_late_inline_helper();
+  }
+
+  virtual void set_process_result(bool v) {
+    _process_result = v;
+  }
+
+  virtual void process_result(GraphKit& kit) {
+    if (!_process_result) {
+      return;
+    }
+    // The call for ScopedValue.get() was just inlined. The code here pattern matches the resulting subgraph. To make it
+    // easier:
+    // - the slow path call to slowGet() is not inlined. If heuristics decided it should be, it was enqueued for late
+    // inlining which will happen later.
+    // - The call to Thread.scopedValueCache() is not inlined either.
+    // The pattern matching here starts from the current control (end of inlining) and looks for the call for
+    // Thread.scopedValueCache() which acts as a marker for the beginning of the subgraph of interest. In the process a
+    // number of checks from the java code of ScopedValue.get() are expected to be encountered. They are recorded:
+    assert(method()->intrinsic_id() == vmIntrinsics::_SVget, "");
+    Compile* C = Compile::current();
+    CallNode* scoped_value_cache = nullptr; // call to Thread.scopedValueCache()
+    IfNode* get_cache_iff = nullptr; // test that scopedValueCache() is not null
+    IfNode* get_first_iff = nullptr; // test for a hit in the cache with first hash
+    IfNode* get_second_iff = nullptr; // test for a hit in the cache with second hash
+    Node* first_index = nullptr; // index in the cache for first hash
+    Node* second_index = nullptr; // index in the cache for second hash
+    CallStaticJavaNode* slow_call = nullptr; // slowGet() call if any
+    {
+      ResourceMark rm;
+      Unique_Node_List wq;
+      wq.push(kit.control());
+      for (uint i = 0; i < wq.size(); ++i) {
+        Node* c = wq.at(i);
```

Comment:
Could we assert that these are all CFG nodes? And give it a more expressive name?